### PR TITLE
feat(fast-components-react-base): add callback to horizontal overflow to return and object that informs scroll start and end

### PR DESCRIPTION
**Pull request by nalogiudice**
Closes [#782](https://github.com/Microsoft/fast-dna/issues/782)
<body>

<footer>
 

For [details on formatting](https://github.com/Microsoft/fast-dna/wiki/pull-request-guidance) pull requests.

### CODE REVIEW

**Comment by nalogiudice** [comment]
@chrisdholt I believe code climate issues are non-issues.

File: packages/fast-components-react-base/src/horizontal-overflow/README.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # Horizontal overflow
 The *horizontal overflow* component accepts items as children. A child can have an optional *slot* property which is either "next" or "previous" which can be used to allow the next/previous click actions. *Horizontal overflow* uses native scrolling behavior.
 
+### Usage guidance
+The *horizontal overflow* component always starts at the start of the overflow contianer &mdash; left side in LTR and right side in RTL.
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/horizontal-overflow/README.md]: Maybe change the first sentence to "The Horizontal overflow component always *initializes*"

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 2]
```diff
@@ -76,6 +90,62 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+    /**
+     * Callback on scroll change
+     */
+    public onScrollChange = (): IScrollObject => {
+        let scrollObject: IScrollObject;
+        const distanceRemaining: number =
+            this.horizontalOverflowItemsRef.current.scrollWidth - this.horizontalOverflowItemsRef.current.scrollLeft;
+
+        if (this.props.onScrollChange) {
+            this.props.onScrollChange();
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: I think you want to pass the `scrollObject` as a parameter to `this.props.onScrollChange` - otherwise an implementer cant know anything about the nature of the scroll change.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 3]
```diff
@@ -76,6 +90,62 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+        }
+
+        if (this.horizontalOverflowItemsRef.current.scrollLeft === 0) {
+            this.getLTR() === Direction.ltr ?
+            scrollObject = {
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: I don't think you want to assign different values here based on rtl - I would expect "start" to be the left edge in ltr and the right edge in rtl. This implementation would require me to check for "start" in ltr and "end" in rtl when trying to determine if I'm at the beginning of the overflow.

File: packages/fast-components-react-base/package.json [hunk 4]
```diff
@@ -103,6 +104,7 @@
     "@microsoft/fast-web-utilities": "^1.9.2",
     "exenv-es6": "^1.0.0",
     "lodash-es": "^4.0.0",
+    "raf-polyfill": "^1.0.0",
```
Comment (nicholasrice) [inline packages/fast-components-react-base/package.json]: I don't think we need `raf-polyfill` as a peer dependency - just a dev dependency because our shipped code doesn't actually reference it.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 5]
```diff
@@ -85,6 +155,18 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+    /**
+     * Request's an animation frame if there are currently no open animation frame requests
+     */
+    private requestFrame = (): void => {
+        if (this.openRequestAnimationFrame) {
+            return;
+        }
+
+        this.openRequestAnimationFrame = true;
+        window.requestAnimationFrame(this.onScrollChange);
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: You might want to look at https://github.com/wuct/raf-throttle instead of manually setting off the rafThrottle - we can then use that package anywhere we need to use raf instead of implementing that logic in every component.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 6]
```diff
@@ -105,6 +187,22 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+        return (
+            <div className={this.props.managedClasses.horizontalOverflow_previous} onClick={this.handlePreviousClick}>
+                {this.getChildBySlot(ButtonDirection.previous)}
+            </div>
+        );
+    }
+
+    private renderNextButton(): JSX.Element {
+        return (
+            <div className={this.props.managedClasses.horizontalOverflow_next} onClick={this.handleNextClick}>
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: I think I would prefer to see these buttons not abstracted to a function - the function doesn't perform any logic and reading the render function now makes me track down what exactly `renderNextButton` does. I think it's more simple to just keep this in the render function.

File: packages/fast-components-react-base/src/horizontal-overflow/README.md (hunk 1 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/README.md]: Sure.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 6 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: This was actually part of another task @janechu for guidance

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 3 above)
Comment (nicholasrice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: Oops - was reading this wrong. Please disregard.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 5 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: Yep, done.

File: packages/fast-components-react-base/package.json [hunk 7]
```diff
@@ -81,6 +81,8 @@
     "jest": "^22.4.2",
     "jsdom-global": "^3.0.2",
     "lodash-es": "^4.0.0",
+    "raf-polyfill": "^1.0.0",
+    "raf-throttle": "^2.0.3",
```
[the remark this answers is not in the document]
Comment (nalogiudice) [reply packages/fast-components-react-base/package.json]: added

File: packages/fast-components-react-base/src/horizontal-overflow/README.md [hunk 8]
```diff
@@ -1,5 +1,8 @@
 # Horizontal overflow
 The *horizontal overflow* component accepts items as children. A child can have an optional *slot* property which is either "next" or "previous" which can be used to allow the next/previous click actions. *Horizontal overflow* uses native scrolling behavior.
 
+### Usage guidance
+The *horizontal overflow* component always initializes at the start of the overflow contianer &mdash; left side in LTR and right side in RTL.
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/README.md]: nit: container, but I would actually change this to content

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.props.ts [hunk 9]
```diff
@@ -1,6 +1,18 @@
...
+ * Scroll interface for consumers
+ * 'start' is when the horizontal overflow scroll is all the way left in LTR (all the way right in RTL)
+ * 'end' is when the horizontal overflow scroll is all the right in LTR (all the way left in RTL)
+ */
+export interface IScrollChange {
+    start: boolean;
+    end: boolean;
+}
+
+export type onScrollChange = (scrollObject: IScrollChange) => void;
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.props.ts]: I believe our naming convention means this should be pascal case.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 10]
```diff
@@ -76,6 +90,43 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+
+        if (this.horizontalOverflowItemsRef.current.scrollLeft === 0) {
+            scrollObject = { start: isLtr, end: !isLtr };
+        } else if (distanceRemaining === this.horizontalOverflowItemsRef.current.clientWidth) {
+            scrollObject = { start: !isLtr, end: isLtr };
+        } else {
+            scrollObject = { start: false, end: false };
+        }
+
+        if (this.props.onScrollChange) {
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: This should be a check at the beginning, if the onScrollChange is not present just return.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 11]
```diff
@@ -76,6 +90,43 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+    public componentWillUnmount(): void {
+        if (canUseDOM() && this.horizontalOverflowItemsRef.current) {
+            this.horizontalOverflowItemsRef.current.removeEventListener("scroll", this.throttled);
+        }
+    }
+
+    /**
+     * Callback on scroll change
+     */
+    public onScrollChange = (): IScrollChange => {
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: why is this public?

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 12]
```diff
@@ -76,6 +90,43 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+        let scrollObject: IScrollChange;
+        const isLtr: boolean = this.getLTR() === Direction.ltr;
+        const distanceRemaining: number =
+            this.horizontalOverflowItemsRef.current.scrollWidth - this.horizontalOverflowItemsRef.current.scrollLeft;
+
+        if (this.horizontalOverflowItemsRef.current.scrollLeft === 0) {
+            scrollObject = { start: isLtr, end: !isLtr };
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: you can just execute the props.onScrollChange

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx [hunk 13]
```diff
@@ -76,6 +90,43 @@ class HorizontalOverflow extends Foundation<HorizontalOverflowProps,  React.AllH
...
+            scrollObject = { start: !isLtr, end: isLtr };
+        } else {
+            scrollObject = { start: false, end: false };
+        }
+
+        if (this.props.onScrollChange) {
+            this.props.onScrollChange(scrollObject);
+        }
+
+        return scrollObject;
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: why does this return anything?

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 10 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: How would we pass them the scroll object with data if it is at the beginning?

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 11 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: So consumers can access it...?

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 13 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: The entire meetings conclusion said we would return an object with scroll data to the consumer.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 12 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: Not sure I understand based on where this comment is

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.props.ts (hunk 9 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.props.ts]: Do you mean `OnScrollChange`?

File: packages/fast-components-react-base/src/horizontal-overflow/README.md (hunk 8 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/README.md]: Sure.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 13 above)
Comment (janechu) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: But that is what you are doing with `this.props.onScrollChange`

Comment (janechu) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: The return value at the end of this does nothing.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 12 above)
Comment (janechu) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: Instead of saving this as a variable you can just execute the this.props.onScrollChange({...data})

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 11 above)
Comment (janechu) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: That is what the callback is for.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.props.ts (hunk 9 above)
Comment (janechu) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.props.ts]: Yep

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 11 above)
Comment (nicholasrice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: To put it more illustratively - it's not really the React convention to expose public methods for consumers to call. To do that, you would have to use a `ref` to obtain a reference to the component and imperatively call a function. Also, publicly calling this function would do a consumer no good because this is a function that needs to be called when the scroll changes - not arbitrarily by the consumer.

Instead, it should be a private function used by the class itself and consumers should register a callback (`this.props.onScrollChage`) to handle updates.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 10 above)
Comment (nicholasrice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: Her point is that this shouldn't return anything - consumers get notified via a registered callback. Because we only care about this data if a callback is registered, we can check for the existence of the callback before doing any work. If the callback doesn't exist, we know we don't need to do any work and we can just exit because the point of this function is to call a consumer-registered callback.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 11 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: I guess I am confused - I thought the function had to be public to access - I will change to private.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 12 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: @janechu are you saying like this?
```
        if (this.horizontalOverflowItemsRef.current.scrollLeft === 0) {
            this.props.onScrollChange({ start: isLtr, end: !isLtr });
        } else if (distanceRemaining === this.horizontalOverflowItemsRef.current.clientWidth) {
            this.props.onScrollChange({ start: !isLtr, end: isLtr });
        } else {
            this.props.onScrollChange({ start: false, end: false });
        }
```

Comment (nicholasrice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: yep - that should work

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 11 above)
Comment (nicholasrice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: That's the point - no one should be accessing it.

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 12 above)
Comment (janechu) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: @nalogiudice yep

Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: Ok

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx (hunk 13 above)
Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.tsx]: Got it

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.spec.tsx [hunk 14]
```diff
@@ -274,7 +274,7 @@ describe("horizontal overflow", (): void => {
             renderedWithImages.instance()[
                 "onScrollChange"
             ]()
-        ).not.toBe(undefined);
+        ).toBe(undefined);
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.spec.tsx]: why is this changed?

**Comment by nalogiudice** [comment]
Follow up task [#823](https://github.com/Microsoft/fast-dna/issues/823)

File: packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.spec.tsx [hunk 15]
```diff
@@ -257,18 +263,38 @@ describe("horizontal overflow", (): void => {
...
-    test("should have an `onLoad` method", () => {
+    test("should unmount appropriately by removing event listener", () => {
         const renderedWithImages: any = mount(
             <HorizontalOverflow managedClasses={managedClasses}>
                 {imageSet1}
             </HorizontalOverflow>
         );
 
+        renderedWithImages.unmount();
+        expect(renderedWithImages.removeEventListener).toBe(undefined);
```
Comment (janechu) [inline packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.spec.tsx]: This does not do anything, I suggest you remove this test.

Comment (nalogiudice) [reply packages/fast-components-react-base/src/horizontal-overflow/horizontal-overflow.spec.tsx]: resolved
